File: serveur/bibliotheque.js
// La bibliothèque accepte l'ancien tableau `etat/books.json` et le futur
// dossier `etat/books/`. Le dossier ne fait foi que lorsque `_ordre.json`
// existe : une copie partielle ne peut donc pas s'activer par accident.
const fs = require("fs");
const path = require("path");

const ID = /^[a-z0-9][a-z0-9._-]*$/;

function emplacements(racine) {
  const dossier = path.join(racine, "etat", "books");
  return {
    dossier,
    manifeste: path.join(dossier, "_ordre.json"),
    monolithe: path.join(racine, "etat", "books.json"),
  };
}

function ordre(manifeste) {
  const ids = JSON.parse(fs.readFileSync(manifeste, "utf-8"));
  if (!Array.isArray(ids) || ids.some((id) => typeof id !== "string"))
    throw new Error("books/_ordre.json doit être une liste d'identifiants");
  if (new Set(ids).size !== ids.length)
    throw new Error("books/_ordre.json contient un identifiant en double");
  const mauvais = ids.find((id) => !ID.test(id));
  if (mauvais) throw new Error("identifiant impropre à un nom de fichier : " + mauvais);
  return ids;
}

function cheminsSource(racine) {
  const e = emplacements(racine);
  if (!fs.existsSync(e.manifeste)) return [e.monolithe];
  return [e.manifeste].concat(ordre(e.manifeste)
    .map((id) => path.join(e.dossier, id + ".json")));
}

function charger(racine) {
  const e = emplacements(racine);
  if (!fs.existsSync(e.manifeste)) {
    const livres = JSON.parse(fs.readFileSync(e.monolithe, "utf-8"));
    if (!Array.isArray(livres)) throw new Error("etat/books.json doit porter une liste");
    return livres;
  }
  return ordre(e.manifeste).map((id) => {
    const fichier = path.join(e.dossier, id + ".json");
    if (!fs.existsSync(fichier)) throw new Error("volume absent : books/" + id + ".json");
    const livre = JSON.parse(fs.readFileSync(fichier, "utf-8"));
    if (!livre || Array.isArray(livre) || typeof livre !== "object")
      throw new Error("books/" + id + ".json doit porter un objet");
    if (livre.id !== id)
      throw new Error("books/" + id + ".json porte l'id " + JSON.stringify(livre.id));
    return livre;
  });
}

function indexer(livres) {
  if (!Array.isArray(livres)) throw new Error("la bibliothèque doit porter une liste");
  const ids = [];
  const parId = new Map();
  livres.forEach((livre) => {
    const id = livre && livre.id;
    if (!ID.test(String(id || "")))
      throw new Error("chaque volume doit porter un id utilisable comme fichier");
    if (parId.has(id)) throw new Error("identifiant de volume en double : " + id);
    ids.push(id);
    parId.set(id, livre);
  });
  return { ids, parId };
}

function egaux(a, b) { return JSON.stringify(a) === JSON.stringify(b); }

function ecrireAtomique(fichier, valeur) {
  fs.mkdirSync(path.dirname(fichier), { recursive: true });
  // Un nom fixe (`volume.json.tmp`) permettait à deux processus de remplir le
  // même temporaire avant que le contrôle optimiste tranche. Le répertoire
  // unique reste sur le même disque, donc le renommage final reste atomique.
  const chantier = fs.mkdtempSync(path.join(path.dirname(fichier), ".bibliotheque-"));
  const temporaire = path.join(chantier, path.basename(fichier) + ".tmp");
  try {
    fs.writeFileSync(temporaire, JSON.stringify(valeur, null, 1), "utf-8");
    fs.renameSync(temporaire, fichier);
  } finally {
    fs.rmSync(chantier, { recursive: true, force: true });
  }
}

function ouvrir(racine) {
  const livres = charger(racine);
  let avant = JSON.parse(JSON.stringify(livres));
  return {
    livres,
    sauver() {
      const e = emplacements(racine);
      const a = indexer(avant);
      const v = indexer(livres);
      const courantsListe = charger(racine);
      const c = indexer(courantsListe);
      if (!fs.existsSync(e.manifeste)) {
        if (!egaux(courantsListe, avant))
          throw new Error("etat/books.json a changé depuis la lecture — rien écrit");
        ecrireAtomique(e.monolithe, livres);
        avant = JSON.parse(JSON.stringify(livres));
        return;
      }
      const tous = new Set([].concat(a.ids, v.ids));
      const touches = Array.from(tous).filter((id) => !egaux(a.parId.get(id), v.parId.get(id)));
      const ordreTouche = !egaux(a.ids, v.ids);
      if (ordreTouche && !egaux(c.ids, a.ids))
        throw new Error("books/_ordre.json a changé depuis la lecture — rien écrit");
      const conflit = touches.find((id) => !egaux(c.parId.get(id), a.parId.get(id)));
      if (conflit) throw new Error("volume modifié depuis la lecture : " + conflit + " — rien écrit");
      v.ids.forEach((id) => {
        if (touches.includes(id)) ecrireAtomique(path.join(e.dossier, id + ".json"), v.parId.get(id));
      });
      if (ordreTouche) ecrireAtomique(e.manifeste, v.ids);
      a.ids.filter((id) => !v.parId.has(id))
        .forEach((id) => fs.unlinkSync(path.join(e.dossier, id + ".json")));
      avant = JSON.parse(JSON.stringify(livres));
    },
  };
}

module.exports = { charger, cheminsSource, ouvrir };

File: scripts/bibliotheque.py
# -*- coding: utf-8 -*-
"""Stockage commun des livres, compatible avec le monolithe et les volumes.

Le dossier n'est actif que si ``etat/books/_ordre.json`` existe. Cela permet
de migrer les lecteurs avant les écrivains sans jamais deviner quelle copie
fait foi. Une bibliothèque activée est stricte : volume absent, identifiant
dupliqué ou fichier dont l'id ne correspond pas à son nom font échouer la
lecture au lieu de retomber silencieusement sur ``books.json``.
"""
import io
import json
import os
import re


NOM_ID = re.compile(r"^[a-z0-9][a-z0-9._-]*$")
MANIFESTE = "_ordre.json"


class BibliothequeInvalide(ValueError):
    pass


def _chemins(etat):
    dossier = os.path.join(etat, "books")
    manifeste = os.path.join(dossier, MANIFESTE)
    monolithe = os.path.join(etat, "books.json")
    return dossier, manifeste, monolithe


def _lire_json(chemin):
    with io.open(chemin, encoding="utf-8") as f:
        return json.load(f)


def _ordre(manifeste):
    ordre = _lire_json(manifeste)
    if not isinstance(ordre, list) or not all(isinstance(i, str) for i in ordre):
        raise BibliothequeInvalide("books/_ordre.json doit être une liste d'identifiants")
    if len(ordre) != len(set(ordre)):
        raise BibliothequeInvalide("books/_ordre.json contient un identifiant en double")
    mauvais = [i for i in ordre if not NOM_ID.fullmatch(i)]
    if mauvais:
        raise BibliothequeInvalide("identifiant impropre à un nom de fichier : %s" % mauvais[0])
    return ordre


def est_scindee(etat):
    return os.path.isfile(_chemins(etat)[1])


def chemins_source(etat):
    """Tous les fichiers dont la date doit invalider un cache de lecture."""
    dossier, manifeste, monolithe = _chemins(etat)
    if not os.path.isfile(manifeste):
        return [monolithe]
    return [manifeste] + [os.path.join(dossier, i + ".json")
                          for i in _ordre(manifeste)]


def charger(etat):
    dossier, manifeste, monolithe = _chemins(etat)
    if not os.path.isfile(manifeste):
        livres = _lire_json(monolithe)
        if not isinstance(livres, list):
            raise BibliothequeInvalide("etat/books.json doit porter une liste")
        return livres

    livres = []
    for ident in _ordre(manifeste):
        chemin = os.path.join(dossier, ident + ".json")
        try:
            livre = _lire_json(chemin)
        except FileNotFoundError as exc:
            raise BibliothequeInvalide("volume absent : books/%s.json" % ident) from exc
        if not isinstance(livre, dict):
            raise BibliothequeInvalide("books/%s.json doit porter un objet" % ident)
        if livre.get("id") != ident:
            raise BibliothequeInvalide(
                "books/%s.json porte l'id %r" % (ident, livre.get("id")))
        livres.append(livre)
    return livres
